### PR TITLE
Improve logging of output for Breeze commands in CI

### PR DESCRIPTION
**Pull request by potiuk**
This change organizes better and improves overal output of the CI breeze
commands, in the way that it makes it far more useful.

1. No more separated output for running command and it's output in
   CI. They were displayed in separately foldable sections and it
   has been somewhat misleading. Now, the folded section contain both
   the command and its output.

2. The heuristics to determine "short" name of a command has been
   improved and the foldable groups for the command are more similar
   to Github Actions runs (white color foldable) but also it has
   pink automatically rich-highlighted main command that is run.

3. CI groups (foldable components) will no longer overlap. Github
   Actions does not support nested groups, so in case we start
   group in another group, the first group is "closed" automatically.
   With this change, if we start group inside of another group, the
   original group is not closed (but also the new group is not started
   and the nested command is not folded inside the main group.

4. Progress output is much more consitent and nicer in case of
   parallel runs. We have it nicely wrapped in rich rules, showing
   how much time started with progress and single line of most
   recent interesting output from the parallel operation. We have
   "smarter" progress information implemented - showing much more
   useful one-liner truncated and caloured status from each of the
   running threads. Then all the individual outputs are nicer
   summarized in the folded groups after all of the parallel runs
   completed. No more spilling of the individual jobs into the
   progress monitoring.

5. When running the same CI parallel commands locally, the output
   is adjusted to local terminal (no foldable groups and the output
   from parallel jobs is only printed if the job fails by default.

This change has been extracted from #25678 as it grew out of it original
scope.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
This one is extracted from #25678 and it provides a number of improvements in the UI of CI output (see the description). Some screenshots:


Parallel runs with progress:

Screenshot 2022-08-22 at 01 54 01

Series of commands with folded output and command defaults:

Screenshot 2022-08-22 at 01 57 36

Some command unfolded:

Screenshot 2022-08-22 at 01 58 12

Another example:

Screenshot 2022-08-22 at 02 01 07

**Comment by potiuk** [comment]
Right. I have run it with "full tests needed and you can see the result of in "build image" "pull image". in #25678  I am useing it in Helm Tests and @bowrna in #23715 might use it after we merge this one.

**Comment by potiuk** [comment]
cc: @Bowrna - once we merge it, you will be able to re-use what I added here without waiting for #25678 - I am struggling there a bit with stability of cluster creations, so it might take a bit more time to complete. But this one has a much nicer and much more reusable way of running parallel tests and presenting the output.

File: dev/breeze/src/airflow_breeze/utils/parallel.py [hunk 1]
```diff
@@ -88,71 +83,142 @@ def get_last_lines_of_file(file_name: str, num_lines: int = 2) -> Tuple[List[str
...
+        self,
+        regexp: str,
+        lines_to_search: int,
+        regexp_for_joined_line: Optional[str] = None,
+        regexp_to_skip_truncation: Optional[str] = None,
+    ):
+        self.last_good_match: Optional[str] = None
+        self.matcher = re.compile(regexp)
+        self.lines_to_search = lines_to_search
+        self.matcher_for_joined_line = re.compile(regexp_for_joined_line) if regexp_for_joined_line else None
```
Comment (potiuk) [inline dev/breeze/src/airflow_breeze/utils/parallel.py:138]: This one allows to see two last important lines of the output (for example in test output we will see):

image

**Review by josh-fell — APPROVED** [review]
Nice

**Comment by Bowrna** [comment]
> cc: @Bowrna - once we merge it, you will be able to re-use what I added here without waiting for #25678 - I am struggling there a bit with stability of cluster creations, so it might take a bit more time to complete. But this one has a much nicer and much more reusable way of running parallel tests and presenting the output.

I am checking the changes and using it in breeze test part @potiuk

**Comment by potiuk** [comment]
> I am checking the changes and using it in breeze test part @potiuk

Cool. It is a bit more involved - with decorators and such, but It is super versatile. You might also take a look how I've implemented the regexps for Helm Tests in https://github.com/apache/airflow/pull/25678/files#diff-ec0edb35ebfa2f51b6030b54f72eb90b0661eb310b0baea6e9c605f49c15123dR1059 

There are two regexps for tests:
1) matching the "interesting" last line
2) matching the "optionall preceding line"

This is the way how in the output of tests you can see two lines of the test progress (which is crucial to see the percentag e progress of tests) while the tests are running

**Comment by potiuk** [comment]
In K8S tests this is achieved by 

```
K8S_TEST_PROGRESS_REGEXP = r'.*airflow-python-[0-9.]+-v[0-9.].*|^kubernetes_tests/.*'
PERCENT_K8S_TEST_PROGRESS_REGEXP = r'^kubernetes_tests/.*\[[ 0-9%]*\].*'
K8S_SKIP_TRUNCATION_REGEXP = r'^kubernetes_tests/.*'

    with ci_group(f"Running tests for: {combos}"):
        with run_with_pool(
            parallelism=parallelism,
            all_params=combo_titles,
            progress_matcher=GenericRegexpProgressMatcher(
                K8S_TEST_PROGRESS_REGEXP,
                lines_to_search=15,
                regexp_for_joined_line=PERCENT_K8S_TEST_PROGRESS_REGEXP,
                regexp_to_skip_truncation=K8S_SKIP_TRUNCATION_REGEXP,
            ),
        ) as (pool, outputs):
```

**Comment by Bowrna** [comment]
> * matching the "interesting" last line

@potiuk Do you mean `interesting` by one that matches with regex?
I think looking into code in `parallel.py` will give me better understanding of this part. Am i right in thinking that way?

**Comment by potiuk** [comment]
> @potiuk Do you mean interesting by one that matches with regex?

Correct. The thing is that in many cases we are interested to see only some output and it might be intertwined with not-interesting ones. So regexp matches onle those lines that provide some info about the progress..

example here:

```

  Database has been reset
  
  
  Those integrations are disabled:  kerberos mongo redis cassandra openldap trino pinot rabbitmq
  
  
  Fixing ownership of 4147 root owned files on Linux
  
  Fixed ownership of generated files.
  
  
  Running tests tests/core tests/executors tests/jobs tests/models tests/serialization tests/ti_deps tests/utils
  
  
  Starting the tests with those pytest arguments: --verbosity=0 --strict-markers --durations=100 --maxfail=50 --color=yes --junitxml=/files/test_result-Core-postgres.xml --timeouts-order moi --setup-timeout=60 --execution-timeout=60 --teardown-timeout=60 -rfEX --with-db-init --cov=airflow/ --cov-config=.coveragerc --cov-report=xml:/files/coverage-Core-postgres.xml tests/core tests/executors tests/jobs tests/models tests/serialization tests/ti_deps tests/utils
  
  ============================= test session starts ==============================
  platform linux -- Python 3.10.6, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
  rootdir: /opt/airflow, configfile: pytest.ini
  plugins: requests-mock-1.9.3, xdist-2.5.0, timeouts-1.2.1, rerunfailures-9.1.1, instafail-0.4.2, httpx-0.21.0, forked-1.4.0, cov-3.0.0, asyncio-0.19.0, flaky-3.7.0, anyio-3.6.1
  asyncio: mode=strict
  setup timeout: 60.0s, execution timeout: 60.0s, teardown timeout: 60.0s
  collected 1978 items
  
  tests/core/test_config_templates.py ....                                 [  0%]
  tests/core/test_configuration.py ....................................... [  2%]
  ...............................................                          [  4%]
  tests/core/test_core.py ......                                           [  4%]
  tests/core/test_example_dags_system.py sssss                             [  5%]
  tests/core/test_impersonation_tests.py sssss                             [  5%]
  tests/core/test_logging_config.py ...........                            [  5%]
  tests/core/test_providers_manager.py .....................               [  6%]
  tests/core/test_sentry.py ....                                           [  7%]
  tests/core/test_settings.py ..........                                   [  7%]
  tests/core/test_sqlalchemy_config.py ...                                 [  7%]
  tests/core/test_stats.py ....................................            [  9%]
  tests/core/test_templates.py .......                                     [ 10%]
  tests/executors/test_base_executor.py .............                      [ 10%]
  tests/executors/test_celery_executor.py sssss...........ssss             [ 11%]
  tests/executors/test_celery_kubernetes_executor.py ....................  [ 12%]
  tests/executors/test_dask_executor.py ..ss....                           [ 13%]
  tests/executors/test_debug_executor.py ......                            [ 13%]
  tests/executors/test_executor_loader.py .......                          [ 13%]
  tests/executors/test_kubernetes_executor.py ............................ [ 15%]
  ...............                                                          [ 15%]
  tests/executors/test_local_executor.py .....                             [ 16%]
  tests/executors/test_local_kubernetes_executor.py .....                  [ 16%]
  tests/executors/test_sequential_executor.py .                            [ 16%]
  tests/jobs/test_backfill_job.py ..x..................................... [ 18%]
  ....ssssss..                                                             [ 19%]
  tests/jobs/test_base_job.py ........                                     [ 19%]
  tests/jobs/test_local_task_job.py .......s...........s.                  [ 20%]
  tests/jobs/test_scheduler_job.py ...........F........................... [ 22%]
  ...........................................s............................ [ 26%]
  ..............ss.................................                     
```

we are not interested in any of the preparation start, but what we want to show as current progress are literally the two last lines:

```
  ...........................................s............................ [ 26%]
  ..............ss.................................                     
```

Ideally with colour (this example is taken from https://github.com/apache/airflow/runs/7977534034?check_suite_focus=true) and you can see that the % is red indicating that we had some failure before.

The Progress Monitor already handles that. you just need to specify the right regexps:

* matching all the "interesting lines" - in our case those with progress
* matching the previous interesing line (they will be mostly the same but there are some cases in Helm chart where I wanted to seperate them because I am mixing "cluster creation" "cluster deployment' and "test execution" in single output so  sometimes I have one-line output, sometimes 2 lines.
* we need to avoid truncation for those "test" lines - and the regexp for that  in case of regular tests will be the same

So I think the "unit test" will be simpe - just one regexp matching the line with ....  should do. It's a bit tricky though - because the progress line might be quite different - mostly it has "." but sometimes "s, x, f, F ' and few others. So you need to make some smart regexp there to match all of them but not match "uninteresting" lines. You can take the output of some of our tests in CI to test those matches.

> I think looking into code in parallel.py will give me better understanding of this part. Am i right in thinking that way?
Correct. The Parallel implemetnation is the right way to look at

**Pull request merged**